File: intersection2.java
/*
Time Complexity: O(N)
Space Complexity: O(N)
*/
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        HashMap<Integer,Integer> hm=new HashMap<>();
        List<Integer> list=new ArrayList<>();
        
        for(int i:nums1){
            if(hm.containsKey(i)){
                hm.put(i,hm.get(i)+1);
            }
            else{
            hm.put(i,1);
            }
        }
        //System.out.println(hm);
        
        for(int i:nums2){
            if(hm.containsKey(i) && hm.get(i)>0){
                hm.put(i,hm.get(i)-1);
                list.add(i);
            }
        }
        //System.out.println(hm);
        int[] ans=new int[list.size()];
        int j=0;
        for(Integer i:list){
            ans[j]=i;
            j++;
        }
        
        return ans; 
        
    }
}

File: intersection_2.py
# Time Complexity :
# O(N)

# Space Complexity :
# O(N)

# Did this code successfully run on Leetcode :
#Yes

#We store the element of num1 and it's frequency of occurance in a map. We then go through each element in nums2, each time we find that element in the map, we reduce it's frequency until it's 0 and add the element to the return list
#If the element from nums2 does not exist or has a frequency of 0 in the map, we skip it

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        freq_map = {}
        for elem in nums1:
            if elem in freq_map :
                freq_map[elem] += 1
            else :
                freq_map[elem] = 1

        return_result = []
        for elem in nums2 :
            if elem in freq_map and freq_map[elem] > 0 :
                return_result.append(elem)
                freq_map[elem] -= 1

        return return_result
